# 1) Data Cleaning

In [1]:
import pandas as pd
import numpy as np

In [2]:
vac_data = pd.read_csv('country_vaccinations.csv')
vac_data

country iso_code        date  total_vaccinations  people_vaccinated  \
0     Afghanistan      AFG  2021-02-22                 0.0                0.0   
1     Afghanistan      AFG  2021-02-23                 NaN                NaN   
2     Afghanistan      AFG  2021-02-24                 NaN                NaN   
3     Afghanistan      AFG  2021-02-25                 NaN                NaN   
4     Afghanistan      AFG  2021-02-26                 NaN                NaN   
...           ...      ...         ...                 ...                ...   
7332     Zimbabwe      ZWE  2021-03-15             37660.0            37660.0   
7333     Zimbabwe      ZWE  2021-03-16             39550.0            39550.0   
7334     Zimbabwe      ZWE  2021-03-17             39607.0            39607.0   
7335     Zimbabwe      ZWE  2021-03-18             40002.0            40002.0   
7336     Zimbabwe      ZWE  2021-03-19             41687.0            41687.0   

      people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                         NaN                     NaN                 NaN   
1                         NaN                     NaN              1367.0   
2                         NaN                     NaN              1367.0   
3                         NaN                     NaN              1367.0   
4                         NaN                     NaN              1367.0   
...                       ...                     ...                 ...   
7332                      NaN                  1301.0               306.0   
7333                      NaN                  1890.0               541.0   
7334                      NaN                    57.0               529.0   
7335                      NaN                   395.0               569.0   
7336                      NaN                  1685.0               772.0   

      total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                               0.00                           0.00   
1                                NaN                            NaN   
2                                NaN                            NaN   
3                                NaN                            NaN   
4                                NaN                            NaN   
...                              ...                            ...   
7332                            0.25                           0.25   
7333                            0.27                           0.27   
7334                            0.27                           0.27   
7335                            0.27                           0.27   
7336                            0.28                           0.28   

      people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                     NaN                             NaN   
1                                     NaN                            35.0   
2                                     NaN                            35.0   
3                                     NaN                            35.0   
4                                     NaN                            35.0   
...                                   ...                             ...   
7332                                  NaN                            21.0   
7333                                  NaN                            36.0   
7334                                  NaN                            36.0   
7335                                  NaN                            38.0   
7336                                  NaN                            52.0   

                vaccines                source_name  \
0     Oxford/AstraZeneca  Government of Afghanistan   
1     Oxford/AstraZeneca  Government of Afghanistan   
2     Oxford/AstraZeneca  Government of Afghanistan   
3     Oxford/AstraZeneca  Government of Afghanistan   
4     Oxford/AstraZeneca  Government of Afghanistan   
... 

In [3]:
vac_data.tail(60)

country  iso_code        date  total_vaccinations  people_vaccinated  \
7277     Wales  OWID_WLS  2021-02-17            841975.0           822633.0   
7278     Wales  OWID_WLS  2021-02-18            864498.0           839065.0   
7279     Wales  OWID_WLS  2021-02-19            885906.0           853904.0   
7280     Wales  OWID_WLS  2021-02-20            897856.0           860083.0   
7281     Wales  OWID_WLS  2021-02-21            905000.0           862248.0   
7282     Wales  OWID_WLS  2021-02-22            919382.0           869653.0   
7283     Wales  OWID_WLS  2021-02-23            937785.0           878506.0   
7284     Wales  OWID_WLS  2021-02-24            959121.0           889270.0   
7285     Wales  OWID_WLS  2021-02-25            982396.0           902334.0   
7286     Wales  OWID_WLS  2021-02-26           1005389.0           916336.0   
7287     Wales  OWID_WLS  2021-02-27           1020023.0           923615.0   
7288     Wales  OWID_WLS  2021-02-28           1029488.0           925669.0   
7289     Wales  OWID_WLS  2021-03-01           1045201.0           933485.0   
7290     Wales  OWID_WLS  2021-03-02           1066798.0           942017.0   
7291     Wales  OWID_WLS  2021-03-03           1090985.0           951540.0   
7292     Wales  OWID_WLS  2021-03-04           1121861.0           967042.0   
7293     Wales  OWID_WLS  2021-03-05           1151582.0           983419.0   
7294     Wales  OWID_WLS  2021-03-06           1171113.0           995059.0   
7295     Wales  OWID_WLS  2021-03-07           1182035.0           998296.0   
7296     Wales  OWID_WLS  2021-03-08           1199421.0          1007391.0   
7297     Wales  OWID_WLS  2021-03-09           1225604.0          1019210.0   
7298     Wales  OWID_WLS  2021-03-10           1256043.0          1034141.0   
7299     Wales  OWID_WLS  2021-03-11           1294144.0          1056787.0   
7300     Wales  OWID_WLS  2021-03-12           1334355.0          1084329.0   
7301     Wales  OWID_WLS  2021-03-13           1370896.0          1113498.0   
7302     Wales  OWID_WLS  2021-03-14           1387186.0          1122931.0   
7303     Wales  OWID_WLS  2021-03-15           1412849.0          1139866.0   
7304     Wales  OWID_WLS  2021-03-16           1446436.0          1157251.0   
7305     Wales  OWID_WLS  2021-03-17           1484566.0          1180155.0   
7306     Wales  OWID_WLS  2021-03-18           1523077.0          1204101.0   
7307  Zimbabwe       ZWE  2021-02-18                 0.0                0.0   
7308  Zimbabwe       ZWE  2021-02-19                 NaN                NaN   
7309  Zimbabwe       ZWE  2021-02-20                 NaN                NaN   
7310  Zimbabwe       ZWE  2021-02-21                 NaN                NaN   
7311  Zimbabwe       ZWE  2021-02-22              1314.0             1314.0   
7312  Zimbabwe       ZWE  2021-02-23              4041.0             4041.0   
7313  Zimbabwe       ZWE  2021-02-24              7872.0             7872.0   
7314  Zimbabwe       ZWE  2021-02-25             11007.0            11007.0   
7315  Zimbabwe       ZWE  2021-02-26             12579.0            12579.0   
7316  Zimbabwe       ZWE  2021-02-27             15705.0            15705.0   
7317  Zimbabwe       ZWE  2021-02-28             18843.0            18843.0   
7318  Zimbabwe       ZWE  2021-03-01             21456.0            21456.0   
7319  Zimbabwe       ZWE  2021-03-02             25077.0            25077.0   
7320  Zimbabwe       ZWE  2021-03-03             27970.0            27970.0   
7321  Zimbabwe       ZWE  2021-03-04             30658.0            30658.0   
7322  Zimbabwe       ZWE  2021-03-05             31325.0            31325.0   
7323  Zimbabwe       ZWE  2021-03-06             32014.0            32014.0   
7324  Zimbabwe       ZWE  2021-03-07             32240.0            32240.0   
7325  Zimbabwe       ZWE  2021-03-08             35518.0            35518.0   
7326  Zimbabwe       ZWE  2021-03-09             35761.0 

In [4]:
vac_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7337 entries, 0 to 7336
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   country                              7337 non-null   object 
 1   iso_code                             7337 non-null   object 
 2   date                                 7337 non-null   object 
 3   total_vaccinations                   4552 non-null   float64
 4   people_vaccinated                    4053 non-null   float64
 5   people_fully_vaccinated              2749 non-null   float64
 6   daily_vaccinations_raw               3830 non-null   float64
 7   daily_vaccinations                   7150 non-null   float64
 8   total_vaccinations_per_hundred       4552 non-null   float64
 9   people_vaccinated_per_hundred        4053 non-null   float64
 10  people_fully_vaccinated_per_hundred  2749 non-null   float64
 11  daily_vaccinations_per_million

In [5]:
# Drop all columns that we don't want
vac_data.drop(['iso_code', 'daily_vaccinations_per_million', 'source_name', 'source_website','daily_vaccinations'], axis=1, inplace=True)

In [6]:
vac_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7337 entries, 0 to 7336
Data columns (total 10 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   country                              7337 non-null   object 
 1   date                                 7337 non-null   object 
 2   total_vaccinations                   4552 non-null   float64
 3   people_vaccinated                    4053 non-null   float64
 4   people_fully_vaccinated              2749 non-null   float64
 5   daily_vaccinations_raw               3830 non-null   float64
 6   total_vaccinations_per_hundred       4552 non-null   float64
 7   people_vaccinated_per_hundred        4053 non-null   float64
 8   people_fully_vaccinated_per_hundred  2749 non-null   float64
 9   vaccines                             7337 non-null   object 
dtypes: float64(7), object(3)
memory usage: 573.3+ KB


In [7]:
pd.set_option('display.max_rows', 100)
vac_data

country        date  total_vaccinations  people_vaccinated  \
0     Afghanistan  2021-02-22                 0.0                0.0   
1     Afghanistan  2021-02-23                 NaN                NaN   
2     Afghanistan  2021-02-24                 NaN                NaN   
3     Afghanistan  2021-02-25                 NaN                NaN   
4     Afghanistan  2021-02-26                 NaN                NaN   
...           ...         ...                 ...                ...   
7332     Zimbabwe  2021-03-15             37660.0            37660.0   
7333     Zimbabwe  2021-03-16             39550.0            39550.0   
7334     Zimbabwe  2021-03-17             39607.0            39607.0   
7335     Zimbabwe  2021-03-18             40002.0            40002.0   
7336     Zimbabwe  2021-03-19             41687.0            41687.0   

      people_fully_vaccinated  daily_vaccinations_raw  \
0                         NaN                     NaN   
1                         NaN                     NaN   
2                         NaN                     NaN   
3                         NaN                     NaN   
4                         NaN                     NaN   
...                       ...                     ...   
7332                      NaN                  1301.0   
7333                      NaN                  1890.0   
7334                      NaN                    57.0   
7335                      NaN                   395.0   
7336                      NaN                  1685.0   

      total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                               0.00                           0.00   
1                                NaN                            NaN   
2                                NaN                            NaN   
3                                NaN                            NaN   
4                                NaN                            NaN   
...                              ...                            ...   
7332                            0.25                           0.25   
7333                            0.27                           0.27   
7334                            0.27                           0.27   
7335                            0.27                           0.27   
7336                            0.28                           0.28   

      people_fully_vaccinated_per_hundred            vaccines  
0                                     NaN  Oxford/AstraZeneca  
1                                     NaN  Oxford/AstraZeneca  
2                                     NaN  Oxford/AstraZeneca  
3                                     NaN  Oxford/AstraZeneca  
4                                     NaN  Oxford/AstraZeneca  
...                                   ...                 ...  
7332                                  NaN   Sinopharm/Beijing  
7333                                  NaN   Sinopharm/Beijing  
7334                                  NaN   Sinopharm/Beijing  
7335                                  NaN   Sinopharm/Beijing  
7336                                  NaN   Sinopharm/Beijing  

[7337 rows x 10 columns]

### Rename columns

In [8]:
vac_data.rename(columns = {'people_vaccinated':'people_only_1st_dose',
                           'people_fully_vaccinated':'people_2nd_dose',
                           'daily_vaccinations_raw':'daily_vaccinations',
                           'total_vaccinations_per_hundred':'%_people_vaccinated',
                           'people_vaccinated_per_hundred':'%_people_only_1st_dose',
                           'people_fully_vaccinated_per_hundred':'%_people_2nd_dose'}, inplace=True)

In [9]:
vac_data

country        date  total_vaccinations  people_only_1st_dose  \
0     Afghanistan  2021-02-22                 0.0                   0.0   
1     Afghanistan  2021-02-23                 NaN                   NaN   
2     Afghanistan  2021-02-24                 NaN                   NaN   
3     Afghanistan  2021-02-25                 NaN                   NaN   
4     Afghanistan  2021-02-26                 NaN                   NaN   
...           ...         ...                 ...                   ...   
7332     Zimbabwe  2021-03-15             37660.0               37660.0   
7333     Zimbabwe  2021-03-16             39550.0               39550.0   
7334     Zimbabwe  2021-03-17             39607.0               39607.0   
7335     Zimbabwe  2021-03-18             40002.0               40002.0   
7336     Zimbabwe  2021-03-19             41687.0               41687.0   

      people_2nd_dose  daily_vaccinations  %_people_vaccinated  \
0                 NaN                 NaN                 0.00   
1                 NaN                 NaN                  NaN   
2                 NaN                 NaN                  NaN   
3                 NaN                 NaN                  NaN   
4                 NaN                 NaN                  NaN   
...               ...                 ...                  ...   
7332              NaN              1301.0                 0.25   
7333              NaN              1890.0                 0.27   
7334              NaN                57.0                 0.27   
7335              NaN               395.0                 0.27   
7336              NaN              1685.0                 0.28   

      %_people_only_1st_dose  %_people_2nd_dose            vaccines  
0                       0.00                NaN  Oxford/AstraZeneca  
1                        NaN                NaN  Oxford/AstraZeneca  
2                        NaN                NaN  Oxford/AstraZeneca  
3                        NaN                NaN  Oxford/AstraZeneca  
4                        NaN                NaN  Oxford/AstraZeneca  
...                      ...                ...                 ...  
7332                    0.25                NaN   Sinopharm/Beijing  
7333                    0.27                NaN   Sinopharm/Beijing  
7334                    0.27                NaN   Sinopharm/Beijing  
7335                    0.27                NaN   Sinopharm/Beijing  
7336                    0.28                NaN   Sinopharm/Beijing  

[7337 rows x 10 columns]

In [10]:
vac_data.shape

(7337, 10)

In [11]:
vac_data.country.describe()

count       7337
unique       150
top       Canada
freq          96
Name: country, dtype: object

    See for missing values

Percentage of missing values

In [12]:
total_mis_val = vac_data.isnull().sum().sum()
total_mis_val

24821

In [13]:
total_cells = np.product(vac_data.shape)
total_cells

73370

In [14]:
per_missing_val = (total_mis_val / total_cells) * 100
per_missing_val

33.82990323020308

>34 % of the data from our dataset is empty